File: 0000-0999/0084.LargestRectangleinHistogram.py
class Solution(object):
    def largestRectangleArea(self, heights):
        """
        :type heights: List[int]
        :rtype: int
        """
        ans = 0
        heights.append(0)
        stack = [-1]
        for i in xrange(len(heights)):
            while heights[i] < heights[stack[-1]]:
                h = heights[stack.pop()]
                w = i - stack[-1] - 1
                ans = max(ans, h*w)
            stack.append(i)
        return ans

        ''' # TLE
        ans = 0
        length = len(heights)
        for i, val in enumerate(heights):
            left, right = i-1, i+1
            while left >= 0 and heights[left] >= val:
                left -= 1
            while right < length and heights[right] >= val:
                right += 1
            ans = max(ans, val*(right-left-1))
        return ans
        '''

File: 0000-0999/0084.LargestRectangleinHistogram.go
func largestRectangleArea(heights []int) int {
    heights = append(heights, 0)
    stack := []int{-1}
    ans := 0
    for end, h := range heights {
        for len(stack) > 1 && h < heights[stack[len(stack)-1]] {
            // use stack[len(stack)-2]+1 instead of stack[len(stack)-1] to deal with duplicated values: ex [2,1,2]
            height := heights[stack[len(stack)-1]]
            start := stack[len(stack)-2] + 1
            ans = max(ans, height * (end-start))
            stack = stack[:len(stack)-1]
        }
        stack = append(stack, end)
    }
    return ans
}

func max(x, y int) int {
    if x > y {
        return x
    }
    return y
}